base/applications/rapps/asyncinet.cpp

```diff
@@ -0,0 +1,258 @@
+                        // so store the handle first, then close them
+                        HINTERNET hInetFile = AsyncInet->hInetFile;
+                        HINTERNET hInternet = AsyncInet->hInternet;
+                        if (hInetFile)
+                        {
+                            InternetCloseHandle(hInetFile);
+                        }
+                        if (hInternet)
+                        {
+                            InternetCloseHandle(hInternet);
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        // read completed immediately.
+                        AsyncInetPerformCallback(AsyncInet, ASYNCINET_DATA, (WPARAM)(AsyncInet->ReadBuffer), (LPARAM)(AsyncInet->BytesRead));
+                    }
+                }
+                else
+                {
+                    DWORD dwError;
+                    if ((dwError = GetLastError()) == ERROR_IO_PENDING)
+                    {
+                        // performing asynchronous IO, everything OK.
+                        break;
+                    }
+                    else
+                    {
+                        //something went wrong
+                        if (dwError == ERROR_INVALID_HANDLE)
+                        {
+                            // That's most likely not an error. handle got closed when canceling may lead to this
+                            break;
+                        }
+                        if (AsyncInet->Callback)
+                        {
+                            AsyncInet->Callback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)dwError, AsyncInet->Extension);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        break;
+        case ERROR_INVALID_HANDLE:
+            if (AsyncInet->bIsCancelled)
+            {
+                // this is most likely not an error. when canceling, using closed/invalid handle may lead to this
+                break;
+            }
+            // fall-down to handle error
+        default:
+            // something went wrong
+            AsyncInetPerformCallback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)(AsyncResult->dwError));
+            break;
+        }
+        break;
+    }
+
+    }
+    return;
+}
+
+
+pASYNCINET AsyncInetDownload(LPCWSTR lpszAgent,
+    DWORD   dwAccessType,
+    LPCWSTR lpszProxy,
+    LPCWSTR lpszProxyBypass,
+    LPCWSTR lpszUrl,
+    BOOL bAllowCache,
+    ASYNCINET_CALLBACK Callback,
+    VOID* Extension
+    )
+{
+    pASYNCINET AsyncInet = NULL;
+    BOOL bSuccess = FALSE;
+    DWORD InetOpenUrlFlag = 0;
+    INTERNET_STATUS_CALLBACK OldCallbackFunc;
+
+    AsyncInet = (pASYNCINET)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(ASYNCINET));
+    if (!AsyncInet)goto cleanup;
```

Comment:
At this point, return NULL would be easier.
And it would be nice to include a debug log that this failed, since it would be quite unexpected.

Reply:
ok, I've improved this.